### PR TITLE
feat(auto-container-ids): Allow tracker init to query a property ID on the fly

### PR DESCRIPTION
**Pull request by zacharyhill**
Automatically get/set container ID for merchant.

Still working to add test cases for the following:
- setPropertyId is called instead of passing propertyId into config object (for backwards compatibility)
- check that correct propertyId goes into cart event
- check that track event queue is cleared when id resolves

### CODE REVIEW

File: src/tracker-component.js [hunk 1]
```diff
@@ -254,6 +271,35 @@ const clearExpiredCart = () => {
...
+        if (paramsToPropertyIdUrl) {
+            url = paramsToPropertyIdUrl();
+        } else {
+            url = 'https://paypal.com/tagmanager/containers/xo';
+        }
+        fetch(`${ url }?mrid=${ merchantId }&url=${ encodeURIComponent(currentUrl) }`)
+            .then(res => {
+                if (res.status === 200) {
+                    return res;
+                }
```
Comment (joshbeam) [inline src/tracker-component.js]: Should we handle the case here if `res.status` isn't 200?

File: src/tracker-component.js [hunk 2]
```diff
@@ -343,7 +389,18 @@ export const Tracker = (config? : Config = defaultTrackerConfig) => {
             };
             track(config, 'setUser', { oldUserId: getUserIdCookie() });
         },
-        setPropertyId: (id : string) => {
+        setPropertyId: (id : string, automatic : boolean = false) => {
```
Comment (joshbeam) [inline src/tracker-component.js]: I'm not a fan of the pattern of having a single method that changes behavior based on a flag that's passed in, I think it increases complexity both on development and usage of the method. Would it make sense to just create another method? We can keep `setPropertyId` as it was originally, and then have a separate internal method for the new stuff.

File: src/tracker-component.js [hunk 3]
```diff
@@ -254,6 +272,35 @@ const clearExpiredCart = () => {
...
+        window.fetch(`${ url }?mrid=${ merchantId }&url=${ encodeURIComponent(currentUrl) }`)
+            .then(res => {
+                if (res.status === 200) {
+                    return res;
+                }
+            })
+            .then(r => r.json()).then(container => {
+                window.localStorage.setItem(propertyIdKey, container.id);
+                resolve(container.id);
+            });
```
Comment (songz) [inline src/tracker-component.js]: add a catch

File: src/tracker-component.js (hunk 1 above)
Comment (songz) [reply src/tracker-component.js]: could we TD this to add proper logging?

File: src/tracker-component.js (hunk 3 above)
Comment (zacharyhill) [reply src/tracker-component.js]: Done. Not sure why github isn't marking this as outdated.

File: src/tracker-component.js [hunk 4]
```diff
@@ -246,6 +256,14 @@ const track = <T>(config : Config, trackingType : TrackingType, trackingData : T
     }
 };
 
+const clearTrackQueue = config => {
```
Comment (joshbeam) [inline src/tracker-component.js]: Does this work? Maybe I have the wrong understanding. I tried to replicate it in the console:

```
let foo = [1,2,3,4,5];
const doIt = () => {
  foo.forEach(int => {
     console.log(int) // equivalent to track(whatever)

     foo.shift();
  })

   console.log(foo)
}

doIt()
VM257:4 1
VM257:4 3
VM257:4 5
VM257:9 (2) [4, 5] // the `foo` queue still has some items left in it
```

It looks like it only fires some events, but that some would be left in the tracking queue?

File: src/tracker-component.js [hunk 5]
```diff
@@ -246,6 +255,14 @@ const track = <T>(config : Config, trackingType : TrackingType, trackingData : T
...
+    // eslint-disable-next-line array-callback-return
+    window.trackEventQueue = window.trackEventQueue.filter(eventData => {
```
Comment (joshbeam) [inline src/tracker-component.js]: Thanks for the fix 👍 Can we add a test for this if we don't have one? I think the behavior would've been caught with that

File: src/tracker-component.js (hunk 4 above)
Comment (zacharyhill) [reply src/tracker-component.js]: Good catch. Pushed a fix. Unfortunately that's a test case I couldn't get to work due to not being able to mock anything (which is why it was added to the window). I'm not sure where/when this got messed up, but I added an issue for this repo to hopefully take care of this issue later so we do not have this issue in the future.

**Pull request merged**